
<!DOCTYPE HTML>

<html lang="en">

<head>
  <meta charset="utf-8">
  <title>ActiveSupport&#58;&#58;Rescuable&#58;&#58;ClassMethods - Ruby on Rails 4.1 - W3cubDocs</title>
  
  <meta name="description" content=" Rescue exceptions raised in controller actions. ">
  <meta name="keywords" content="module, activesupport, rescuable, classmethods, -, ruby, on, rails, rails~4.1">
  <meta name="HandheldFriendly" content="True">
  <meta name="MobileOptimized" content="320">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="http://docs.w3cub.com/rails~4.1/activesupport/rescuable/classmethods/">
  <link href="/favicon.png" rel="icon">
  <link type="text/css" rel="stylesheet" href="/assets/application-50364fff564ce3b6327021805f3f00e2957b441cf27f576a7dd4ff63bbc47047.css">
  <script type="text/javascript" src="/assets/application-db64bfd54ceb42be11af7995804cf4902548419ceb79d509b0b7d62c22d98e6f.js"></script>
  <script src="/json/rails~4.1.js"></script>
  
  <script type="text/javascript">
    var _gaq = _gaq || [];
    _gaq.push(['_setAccount', 'UA-71174418-1']);
    _gaq.push(['_trackPageview']);

    (function() {
      var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
      ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
      var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
    })();
  </script>


</head>

<body>
	<div class="_app">
	<header class="_header">
  
  <form class="_search">
    <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
    <a class="_search-clear"></a>
    <div class="_search-tag"></div>
  </form>
  
  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/rails~4.1/" class="_nav-link" title="" style="margin-left:0;">Ruby on Rails 4.1</a></span>
  
  <nav class="_nav">
    <a href="/app/" class="_nav-link ">App</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		<div class="_list">
			
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<div class="_page _rdoc">
				
<h1 id="module-ActiveSupport::Rescuable::ClassMethods" class="module"> module ActiveSupport::Rescuable::ClassMethods </h1>  <section id="5Buntitled-5D" class="documentation-section"> <section id="public-instance-5Buntitled-5D-method-details" class="method-section"> <header> <h3>Public Instance Methods</h3> </header> <div class="method-detail "> <div class="method-heading" id="method-i-rescue_from"> <span class="method-name">rescue_from</span><span class="method-args">(*klasses, &amp;block)</span> <a class="method-click-advice">Show source</a> </div> <div class="method-description"> <p>Rescue exceptions raised in controller actions.</p> <p><code>rescue_from</code> receives a series of exception classes or class names, and a trailing <code>:with</code> option with the name of a method or a Proc object to be called to handle them. Alternatively a block can be given.</p> <p>Handlers that take one argument will be called with the exception, so that the exception can be inspected when dealing with it.</p> <p>Handlers are inherited. They are searched from right to left, from bottom to top, and up the hierarchy. The handler of the first class for which <code>exception.is_a?(klass)</code> holds true is the one invoked, if any.</p> <pre>class ApplicationController &lt; ActionController::Base
  rescue_from User::NotAuthorized, with: :deny_access # self defined exception
  rescue_from ActiveRecord::RecordInvalid, with: :show_errors

  rescue_from 'MyAppError::Base' do |exception|
    render xml: exception, status: 500
  end

  protected
    def deny_access
      ...
    end

    def show_errors(exception)
      exception.record.new_record? ? ...
    end
end</pre> <p>Exceptions raised inside exception handlers are not propagated up.</p> <div class="method-source-code" id="rescue_from-source"> <pre class="ruby" data-language="ruby"># File activesupport/lib/active_support/rescuable.rb, line 51
def rescue_from(*klasses, &amp;block)
  options = klasses.extract_options!

  unless options.has_key?(:with)
    if block_given?
      options[:with] = block
    else
      raise ArgumentError, "Need a handler. Supply an options hash that has a :with key as the last argument."
    end
  end

  klasses.each do |klass|
    key = if klass.is_a?(Class) &amp;&amp; klass &lt;= Exception
      klass.name
    elsif klass.is_a?(String)
      klass
    else
      raise ArgumentError, "#{klass} is neither an Exception nor a String"
    end

    # put the new handler at the end because the list is read in reverse
    self.rescue_handlers += [[key, options[:with]]]
  end
end</pre> </div> </div> </div> </section> </section><div class="_attribution">
  <p class="_attribution-p">
    © 2004–2016 David Heinemeier Hansson<br>Licensed under the MIT License.<br>
    
  </p>
</div>

			</div>
		</div>
	</section>

	</div>
</body>
</html>
